### ISSUE TITLE
macos: specifying static initializer routines in Zig explicitly

### ISSUE THREAD

**Issue by kubkon** [ziglang/zig#13319]
### Zig Version

0.10.0-dev.4572+453a2d2fa

### Steps to Reproduce

After a cool debugging session with @cryptocode, we have discovered a potential footgun to do with `linksection` on macOS. The point of the exercise was to force a function to become a static initializer routine. Static initializers are registered as relocatable pointers in `__DATA,__mod_init_func` section in the relocatable object file.

Take this example which at first sight should do what we want:

```zig
const std = @import("std");

export fn mod_init_func() callconv(.C) void {
    std.debug.print("Loading\n", .{});
}

export const mod_init_func_ptr linksection("__DATA,__mod_init_func") = mod_init_func;

pub fn main() u8 {
    std.debug.print("Main\n", .{});
    return 0;
}
```

### Expected Behavior

```
$ zig run main.zig
Loading
Main

```

### Actual Behavior

Zig fails to generate a `__mod_init_func` pointer entry for `mod_init_func` function. I believe here we might be setting `mod_init_func_ptr` as a `comptime` constant and thus not exporting it, but this is just a guess.

```
$ zig run main.zig
Main

```

If we force pointer via address-of operator on `mod_init_func`, it works fine:

```zig
const std = @import("std");

export fn mod_init_func() callconv(.C) void {
    std.debug.print("Loading\n", .{});
}

export const mod_init_func_ptr linksection("__DATA,__mod_init_func") = &mod_init_func; // note the & op

pub fn main() u8 {
    std.debug.print("Main\n", .{});
    return 0;
}
```

```
$ zig run main.zig
Loading
Main

```

**Comment by JCash** [comment]
When I tested this in my project, it worked fine at first, but then it stopped working.
I've seen this before (In the Defold code), with the linker optimizing away the code, and the global initializer is never invoked.

Our approach to solving it is to trick the linker that we export another function which references the global initializer function.
Example in my test project:
https://github.com/JCash/zig-plugin-test/blob/main/src/main.cpp#L4-L9

**Comment by mikdusan** [comment]
This code attempts to export a couple of things that are non-exportable:

### z0.zig
```zig
export const a: type = @TypeOf(bool);
export const b: u3 = 1;
```

```sh
$ zig build-obj z0.zig
z0.zig:1:1: error: unable to export type 'type'
export const a: type = @TypeOf(bool);
^~~~~~
z0.zig:2:1: error: unable to export type 'u3'
export const b: u3 = 1;
^~~~~~
z0.zig:2:1: note: only integers with power of two bits are extern compatible
```

Perhaps we need an error such as:
```
error: unable to export const type 'fn() callconv(.C) void'
note: use '*const fn() callconv(.C) void' for a function pointer type
```

**Comment by recursivetree** [comment]
I've started to look into this problem, it seems like the `__mod_init_func` section is not emitted without the pointer. However, I started to think that if it requires a function pointer in `__mod_init_func`, it makes sense that you have to explicitly put a function pointer there. As far as I understand it, without the pointer is should put the function code there, which is obviously wrong.

**Comment by kubkon** [comment]
> I've started to look into this problem, it seems like the `__mod_init_func` section is not emitted without the pointer. However, I started to think that if it requires a function pointer in `__mod_init_func`, it makes sense that you have to explicitly put a function pointer there. As far as I understand it, without the pointer is should put the function code there, which is obviously wrong.

Correct, `__mod_init_func` contains relocatable pointers to symbols within `__TEXT,__text` section or equivalent. When `__mod_init_func` is emitted by the compiler in an object file, it usually points via a relocation into `__TEXT,__StaticInit` section which is meant to merged into `__TEXT,__text` by the linker.

**Comment by alexrp** [comment]
Related: https://github.com/ziglang/zig/issues/23574
